### ISSUE TITLE
should we continue to use hypermail?

### ISSUE THREAD

**Issue by gosko** [w3c/mailing-list-archives#8]
W3C has been using [hypermail](https://github.com/hypermail-project/hypermail) to generate its mailing list archives for the past 20 years or so.  It generally works well but has a few drawbacks:
- no way to view an entire thread on a single page.
- message threads are broken at time-based boundaries (e.g. monthly or quarterly)
- no active user/developer community

However, there doesn't seem to be much in the way of software we could use to replace hypermail.  [Hyperkitty](https://hyperkitty.readthedocs.org/en/latest/) is the only option I have seen that looks viable, but switching archive systems would be fairly disruptive to our user community, and hyperkitty likely has its own set of drawbacks that we don't know about.

So I think we should plan to continue using hypermail for the forseeable future, and try to improve some of its thread navigation features using external code.

But I wanted to open this as an issue in case others want to propose we switch to something else, and to have a place to discuss possible alternatives.

**Comment by marcoscaceres** [comment]
We could fork hypermail and at least hack a bit on the HTML output? It's not as if anyone is maintaining it, and maybe the HTML output parts are easy to change. Worth a look.

**Comment by gosko** [comment]
> We could fork hypermail and at least hack a bit on the HTML output? It's not as if anyone is maintaining it, and maybe the HTML output parts are easy to change. Worth a look.

@marcoscaceres W3C has been the main contributor for the past decade or so (thanks primarily to Jose Kahan), and we have managed to avoid forking it so far. We can certainly reconsider that and do a fork if the changes we want would be too disruptive to the main hypermail project.

**Comment by marcoscaceres** [comment]
It would be great if the modifications made their way back upstream.

**Comment by tripu** [comment]
I was just looking around Hypermail. As a newbie in that project, I submitted [one PR and a few suggestions in the form of issues](https://github.com/hypermail-project/hypermail/issues?utf8=%E2%9C%93&q=author%3Atripu) to try to improve its _forkability_ a bit.

**Comment by gosko** [comment]
[This discussion about gmane](http://lwn.net/Articles/695695/)  has some interesting comments about some of its most beloved features. (and one person who really hates hyperkitty)

In particular, a number of people commented how useful it is to be able to create a link to a message based on its msgid, pointing to a threaded view with the specific message highlighted.

**Comment by marcoscaceres** [comment]
Mozilla moved our mailing lists to Google's list-thing. They are really good - particularly for searching, replying, etc. Why don't we just do that?

**Comment by gosko** [comment]
@marcoscaceres a few reasons come to mind: privacy, persistence, decentralization.

Also I expect it would be frustrating not to have any control over how things work, and Google doesn't respond to user feedback.

**Comment by marcoscaceres** [comment]
> @marcoscaceres a few reasons come to mind: privacy, persistence, decentralization.

Not much different to what we have today. Persistence and decentralization can still be handled by keeping a copy of the list archive (assuming that is possible somehow). 

> Also I expect it would be frustrating not to have any control over how things work, and Google doesn't respond to user feedback.

Ok, but compare that to the 20+ year old software we are using today.

**Comment by gosko** [comment]
> Not much different to what we have today.

!? The privacy situation couldn't be any more different, between us hosting it vs. Google.

By persistence I didn't mean losing access to the data, but breaking all the links. I would say one of the best features of our archives (maybe the single best feature) is having the ability to link to something and trust that the link will continue to work 5, 10, or 20 years from now.

Re: decentralization, putting more and more of our data in the hands of a small handful of companies is actively harmful to the health of the Web.

> Ok, but compare that to the 20+ year old software we are using today. 

Hence this project :)

And it hasn't been unmaintained all that time, we have made a number of improvements. It just needs a bit of an update...

**Comment by fantasai** [comment]
Totally agree with @gosko's last comment. We use links to individual messages in the email archives a lot, and being able to go back and make sense of historical discussions is important. This won't be possible except with a system we host ourselves--and if we switch systems, we're still going to need those old links to work.

**Comment by fantasai** [comment]
I'd say, if we were going to continue to use email for issue-tracking, switching to a different system would be worth it if it let us get features that made that more workable. But since we're all switching to GitHub for that, just making hypermail a little bit more usable by addressing the threading issue and adding a handful of minor usability fixes is enough for me.

Fwiw, I'm opposed to switching to some JS-based system, I like the [last-century perf characteristics](https://twitter.com/gravislizard/status/927593460642615296) of the current system. :p

**Comment by tripu** [comment]
@fantasai: I agree about not relying on JS. All usability changes proposed so far are &ldquo;progressive enhancements&rdquo; (controversial term, I know), on top of pages that are functional also without JS (try disabling JS on [better headers](https://w3c.github.io/mailing-list-archives/samples/message-proposal-fields.html), [gravatars](https://w3c.github.io/mailing-list-archives/samples/message-proposal-gravatars.html), [formatting](https://w3c.github.io/mailing-list-archives/samples/message-proposal-markdown.html), etc). If by mistake we did something that breaks when there's no JS, we'll change that!

**Comment by gosko** [comment]
I just discovered another archive system, used by the apache software foundation: [Pony Mail](https://ponymail.incubator.apache.org/). I am not a fan of the styling but some of the features may be interesting.

We still have no plans to move beyond hypermail or invest much time in this project in the near future; just mentioning this here since I'm using this issue to track possible alternatives and look for ideas worth stealing.

**Comment by ericprud** [comment]
+1 to keeping an eye out for other archive systems.

re Pony Mail, I like their [reply rationale](https://ponymail.incubator.apache.org/about.html#rationale) but it doesn't appear to have a tree-threaded UI. While Google has concluded otherwise, I strongly believe that tree-threading is essential. For day-to-day issues, it's possible got get by with linear theading (like this issue), but a navigable tree is essential for e.g. making sure that public comments are fully addressed. (Github and its users invent lots of conventions for splitting off issues to get around this but in my adamant opinion, they're crappy approximations of a tree forced, by tooling limitations.)
